pkg/fqdn/service/service.go

```diff
@@ -102,17 +122,102 @@ type PolicyUpdater interface {
 
 // StreamPolicyState is a bidirectional streaming RPC to subscribe to DNS policies
 // SDP calls this method to subscribe to DNS policies
-// For each stream, we start a goroutine to receive the DNS policies ACKs
+// For each stream, we start a goroutine to receive the DNS policies ACKs for that particular client.
 // The flow of the method is as follows:
-// 1. Add the stream to the map( called by the client i.e SDP)
-// 2. Start a goroutine to receive the DNS policies ACKs for that particular client.
-// 3. Send the current state of the DNS rules to the client (We store the current state fo DNS rules during the endpoint regeneration see UpdatePolicyRulesLocked)
-// 4. Wait for the context to be done
-// Note: this method is left empty on purpose and will be update with the actual implementation in the future PRs for the standalone DNS proxy
+//  1. Add the stream to the local map(called by the client i.e SDP)
+//  2. Start a goroutine to receive the DNS policies ACKs for that particular client.
+//  3. Send the current state of the DNS rules to the client (We store the current state of
+//     DNS rules during the endpoint regeneration see UpdatePolicyRules)
+//  4. Wait for the context to be done and return the error if any
+//  5. Delete the stream from the local map
 func (s *FQDNDataServer) StreamPolicyState(stream pb.FQDNData_StreamPolicyStateServer) error {
-	// This is a temporary implementation to send the current state of the DNS rules to the client and used for testing
-	stream.Send(&pb.PolicyState{RequestId: "test"})
-	return nil
+	streamCtx, cancel := context.WithCancel(stream.Context())
+	s.streams.Store(stream, cancel)
+
+	ackRecvStream := make(chan struct{}, 1)
+	// Start a goroutine to receive the DNS policies ACKs
+	go func() {
+		if err := s.receiveDNSPolicyACKs(stream, ackRecvStream); err != nil {
+			s.log.Error("Error receiving DNS policies ACK", logfields.Error, err)
+			cancel() // Cancel the context to close the stream
+			s.deleteStream(stream)
+			return
+		}
+	}()
+
+	// Wait for the ACK to be received
+	<-ackRecvStream
```

Comment:
What exactly does this do? `receiveDNSPolicyACKs` closes this channel right away so this is essentially only synchronizing that the goroutine got created it happened to run long enough to close this channel. If this is for actually making sure that it has called `stream.Recv` then this is not reliable since the `receiveDNSPolicyACKs` goroutine could've been pre-empted between the `close(ack)` and the `select` statement.

Or is the `close(ack)` misplaced? Should it be `defer close(ack)`?